sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryQueryHelper.java

```diff
@@ -96,105 +96,112 @@ public static TableReference executeQuery(
-            .setCreateDisposition("CREATE_IF_NEEDED")
-            .setWriteDisposition("WRITE_TRUNCATE")
-            .setPriority(priority.name());
-
-    if (kmsKey != null) {
-      queryConfiguration.setDestinationEncryptionConfiguration(
-          new EncryptionConfiguration().setKmsKeyName(kmsKey));
-    }
+      // Step 3: Execute the query. Generate a transient (random) query job ID, because this code
+      // may be retried after the temporary dataset and table have been deleted by a previous
+      // attempt -- in that case, we want to regenerate the temporary dataset and table, and we'll
+      // need a fresh query ID to do that.
+      LOG.info(
+          "Exporting query results into temporary table {} using job {}",
+          queryResultTable,
+          queryJobId);
+
+      JobReference jobReference =
+          new JobReference()
+              .setProjectId(
+                  options.getBigQueryProject() == null
+                      ? options.getProject()
+                      : options.getBigQueryProject())
+              .setLocation(effectiveLocation)
+              .setJobId(queryJobId);
+
+      JobConfigurationQuery queryConfiguration =
+          createBasicQueryConfig(query, flattenResults, useLegacySql)
+              .setAllowLargeResults(true)
+              .setDestinationTable(queryResultTable)
+              .setCreateDisposition("CREATE_IF_NEEDED")
+              .setWriteDisposition("WRITE_TRUNCATE")
+              .setPriority(priority.name());
+
+      if (kmsKey != null) {
+        queryConfiguration.setDestinationEncryptionConfiguration(
+            new EncryptionConfiguration().setKmsKeyName(kmsKey));
+      }
 
-    JobService jobService = bqServices.getJobService(options);
-    jobService.startQueryJob(jobReference, queryConfiguration);
-    Job job = jobService.pollJob(jobReference, JOB_POLL_MAX_RETRIES);
-    if (BigQueryHelpers.parseStatus(job) != Status.SUCCEEDED) {
-      throw new IOException(
-          String.format(
-              "Query job %s failed, status: %s",
-              queryJobId, BigQueryHelpers.statusToPrettyString(job.getStatus())));
-    }
+      JobService jobService = bqServices.getJobService(options);
+      jobService.startQueryJob(jobReference, queryConfiguration);
+      Job job = jobService.pollJob(jobReference, JOB_POLL_MAX_RETRIES);
+      if (BigQueryHelpers.parseStatus(job) != Status.SUCCEEDED) {
+        throw new IOException(
+            String.format(
+                "Query job %s failed, status: %s",
+                queryJobId, BigQueryHelpers.statusToPrettyString(job.getStatus())));
+      }
 
-    LOG.info("Query job {} completed", queryJobId);
-    return queryResultTable;
+      LOG.info("Query job {} completed", queryJobId);
+      return queryResultTable;
+
+    } catch (RuntimeException | IOException | InterruptedException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
```

Comment:
Are the new catch blocks needed?

Reply:
The method throws `InterruptedException` and `IOException` only.  I refer to [BigQueryHelpers.getNumRows](https://github.com/apache/beam/blob/v2.38.0-RC1/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryHelpers.java#L554-L556).

Reply:
But the catch blocks were not there before, right?

Reply:
`DatasetService.close()` throws `Exception` (not `IOException`) and the method `executeQuery` has only `IOException` and `InterruptedException` as throws in the method signature. Without the new catch blocks, `try (DatasetService tableService = bqServices.getDatasetService(options)) { ... }` causes compile error.